### PR TITLE
Collision detection for IMAT and other minor edits

### PR DESCRIPTION
**Pull request by DominicOram**
### Description of work

Added some improvements to collision detection.

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Has the author taken into account the multi-threaded nature of the code?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: CollisionDetection/configurations/config_zoom.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+control_pv = "{}COLLIDE:".format(pv_prefix)
+
+# Define the geometry of the system in mm
+# Coordinate origin at arc centre, with nominal beam height
+# Size is defined x, y, z (x is beam, z is up)
+
+detector_x_size = 1000
+baffle_x_size = 1000
+
+detector = dict(name="Detector", size=(detector_x_size, 1000.0, 1000.0), color=MAGENTA)  # Initial height is halfway across travel
```
Comment (DominicOram) [inline CollisionDetection/configurations/config_zoom.py]: Not sure what is meant by this comment?

File: CollisionDetection/collide.py [hunk 2]
```diff
@@ -9,19 +10,26 @@
...
+def collide(geometries, ignore, collision_func=ode.collide):
+    """
+    Calculates which of the given geometries will collide, ignoring geometries that are specified as ignored.
+    As there are only [(len(geometries)-1)!] combinations, and we don't care about some, there isn't much effort saved
+    by using spaces (which do a quicker estimate of collisions first)
+
+    Args:
+        geometries: A list of GeometryBox objects to check for collisions.
+        ignore: A list of pairs to ignore. Each pair is represented by a list with two entries.
+        collision_func: A callable which takes two geometries as input, and returns True iff they are colliding.
```
Comment (DominicOram) [inline CollisionDetection/collide.py]: iff, typo

File: CollisionDetection/collide.py [hunk 3]
```diff
@@ -9,19 +10,26 @@
...
 
-# This ignores geometries we have said we don't care about
-# As there are only [(len(geometries)-1)!] combinations, and we don't care about some, there isn't much effort saved
-# by using spaces (which do a quicker estimate of collisions first)
-def collide(geometries, ignore):
+def collide(geometries, ignore, collision_func=ode.collide):
+    """
+    Calculates which of the given geometries will collide, ignoring geometries that are specified as ignored.
+    As there are only [(len(geometries)-1)!] combinations, and we don't care about some, there isn't much effort saved
+    by using spaces (which do a quicker estimate of collisions first)
```
Comment (DominicOram) [inline CollisionDetection/collide.py]: what is meant by using spaces?

**Comment by DominicOram** [comment]
Docstrings and tests look like a good start. However:
* Can you write a follow up ticket(s) to catch what's left to do
* The tests aren't in jenkins and the dependencies aren't in our version of python.
* You've changed MOVN to DMOV but haven't inverted any of the logic to check for movement, therefore it thinks it's in constant motion.
* The motors are actually 0501 = Baffle, 0507 = Detector
* I'm not sure the geometry is being calculated correctly. I would expect when both at zero the baffle to have a limit at roughly +5789 and the detector at -5789 (this is probably my fault for poorly explaining the geometry)
* The calculated limits seem to jump around a worrying amount. If I keep the detector stationary I would expect the baffle's limits not to move.
* It seems to take a worrying amount of time to tell there was a collision, this is probably outside of the scope of this.

File: CollisionDetection/collide.py (hunk 2 above)
Comment (Tom-Willemsen) [reply CollisionDetection/collide.py]: https://en.oxforddictionaries.com/definition/iff

Comment (DominicOram) [reply CollisionDetection/collide.py]: Still confusing, sorry

File: CollisionDetection/transform.py [hunk 4]
```diff
@@ -88,103 +105,83 @@ def translate(self, x=0, y=0, z=0, forward=True):
...
 
         return self.matrix
 
-    def to_opengl(self):
-        """
-        Reshape to the format expected by OpenGL
-        """
-        return np.reshape(self.matrix.T, 16)
-
-    # Doesn't get the same result as below!
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply CollisionDetection/transform.py]: Thanks for clarification

File: CollisionDetection/collide.py (hunk 3 above)
Comment (Tom-Willemsen) [reply CollisionDetection/collide.py]: I'm not sure. I'd left the comment as it was, but as it is confusing I will remove it.

**Comment by Tom-Willemsen** [comment]
I've created a follow up ticket at https://github.com/ISISComputingGroup/IBEX/issues/3141 . Other comments addressed.

**Comment by DominicOram** [comment]
As discussed:
- Add requirements wheel to share
- Ensure DMOV logic is correct  in all places

**Pull request merged**